### ISSUE TITLE
#432 is highly incompatible => add additional flag

### ISSUE THREAD

**Issue by cohomology** [PCRE2Project/pcre2#574]
Hi,

You introduced an incompatible change in PCRE2 with #432, which caused many of our unit tests to fail. 

Would it be possible to introduce a compile flag to restore the old behavior? If so, I could supply a patch as well.

Couldn't this affect also many other software products and trigger subtle bugs?

Thanks
Kilian

**Comment by cohomology** [comment]
We are authors of a programming language. We don't write the regex or the subject, but the users of the programming language do. We don't know what regular expressions they write. 

This programming language is compatible along releases. So an old program should run the same as a new program. 

Do you know how other programming languages like PHP consume PCRE2? I guess they also try to be compatible along releases and do not invalidate user code. Do they fork PCRE?

The best and most modern approach would be to use feature flags throughout. Is that an option?

**Comment by cohomology** [comment]
I have seen Perl doesn't seem to mind invalidating user code, see: https://github.com/php/php-src/pull/13413

But this might be one reason very few people still use perl :-(

**Comment by carenas** [comment]
> We are authors of a programming language

Just curious, but who is we and which language?

PCRE2 is used in several and some have their own per language comptibility flag that might come handy on that upcoming patch of yours.

> I also feel that the old behavior was more consistent. But that's just personal taste.

Philip seemed to agree with you, so I am sure something implementing `PCRE2_EXTRA_ALT_CASELESS_BSP` or something more properly named is likely to have no problem getting merged

**Comment by PhilipHazel** [comment]
Indeed, a new option seems like the best way out of this, and that seems like the appropriate name (similar to other option names). As you can see from the comments in #432, the change was made in response to user complaints. We try to please everybody, but sometimes an option is the only way to do it. Awaiting your patch....

**Comment by PhilipHazel** [comment]
Note that the change also fixes a bug in POSIX classes. Previously, [:upper:] (for example) matched only upper case when PCRE2_UCP was set, but matched both cases when PCRE2_UCP was not set. This is because [:upper:] is translated to \p{Lu} when PCRE2_UCP is set. I suppose for consistency [:upper:] should always behave in the same way as \p{Lu}.

**Comment by zherczeg** [comment]
Please understand, that PCRE2 is not funded in any way, so developers spend their free time on it. There are things we do, but spending a lot of effort on minor things is not considered a valuable effort (at least from my side). Adding flags to everything increases the maintenance costs, and I would avoid that. I know people often consider regex as something simple, and doing these requests are simple tasks, but I would be curious how Nick's opinion changed about the project before and after he joined it (from a newcomers perspective).

**Comment by cohomology** [comment]
@carenas The ABAP programming language. It is used to run large enterprise software and has very strict compatibility requirements.
@zherczeg I totally understand. This is why I proposed to supply a patch myself. I would only add feature flags if the behavior of existing regular expressions is changed totally in an incompatible way. This happens very seldom, I guess. Being compatible is a value of itss own and makes PCRE more useful in real-world applications. I agree that adding flags for every new feature is not acceptable for such a project. 

We are still discussing, whether we can change the behavior also in ABAP. Searching \p{Lu} with "i" flag is quite obscure. Probably no one did that.

**Comment by carenas** [comment]
> Searching \p{Lu} with "i" flag is quite obscure.

don't forget also the effect it has in other related properties and even potencial aliases, as pointed out by Philip in the case of `[:upper:]` under UCP mode.

**Comment by NWilson** [comment]
> We are still discussing, whether we can change the behavior also in ABAP. Searching \p{Lu} with "i" flag is quite obscure. Probably no one did that.

We haven't heard from you (which is fine by the way), so I will be proceeding with the next 10.45 release on the assumption that it's OK for us to make this change to behaviour, without a flag to keep the old behaviour. It is mentioned as one of the top items in the release notes.

If this means you can't update to 10.45, don't worry: you or I can always add a flag in 10.46 (or later).
